### ISSUE TITLE
Support or bridge the I2P network via a plugin in Tribler

### ISSUE THREAD

**Issue by devos50** [Tribler/tribler#6737]
As suggested by a user in a [Reddit thread](https://www.reddit.com/r/tribler/comments/s6yz9p/regarding_dht_node_calls_block_issue/):

I2P network is supported or bridged via a plugin in Tribler.

This bridge will help users of Tribler, as new set of peers from I2P will help in uploading/downloading the torrent. Thereby, bypassing the bottleneck of DHT Blocked calls currently faced by Tribler.

![ogb871avggc81](https://user-images.githubusercontent.com/1707075/150326741-c348c5a2-e203-4d78-8327-f6ad9017d078.png)

I am thinking on the line of something that can be seen from above image.

**Comment by absolutep** [comment]
any update on this?

**Comment by absolutep** [comment]
Libtorrent has released new version which supports I2P now = https://github.com/arvidn/libtorrent/releases

Can we have this feature now?

**Comment by xoriole** [comment]
@absolutep The support for libtorrent v2 is not yet available on Tribler so will have to wait before we can have this.

**Comment by teddyrogers** [comment]
@qstokkink with the addition of I2P in Tribler, do we need to install and connect to I2P separately or will it be similar to BiglyBT where it comes included with the installer?

https://github.com/BiglySoftware/BiglyBT/wiki/I2P

**Comment by absolutep** [comment]
> [@qstokkink](https://github.com/qstokkink) with the addition of I2P in Tribler, do we need to install and connect to I2P separately or will it be similar to BiglyBT where it comes included with the installer?
> 
> https://github.com/BiglySoftware/BiglyBT/wiki/I2P

More importantly, will Tribler support I2P DHT or I2P PeX??

**Comment by qstokkink** [comment]
> install and connect to I2P separately

Currently, we allow _connections from_ peers behind I2P. There is an option to route Tribler to the local I2P client, but we don't use this (I'm not sure if this even makes sense, given our own anonymization).

> will Tribler support I2P DHT or I2P PeX?

This is not supported by our [underlying library](https://github.com/arvidn/libtorrent/issues/7408). That means this is a bit more effort than "turn feature on". So, maybe in the future. But, not short term. We would have to first investigate the protocol, see if it fits our implementation, revise our software architecture, etc. etc.

**Comment by teddyrogers** [comment]
> Currently, we allow _connections from_ peers behind I2P. 

Correct me if wrong, Triblers I2P implementation will route I2P traffic over the Tribler network? 

> There is an option to route Tribler to the local I2P client, but we don't use this (I'm not sure if this even makes sense, given our own anonymization).

I agree, it does not make sense to me. I was going to make the recommendation against if that were to be the implementation.

**Comment by qstokkink** [comment]
> Correct me if wrong, Triblers I2P implementation will route I2P traffic over the Tribler network?

It depends on how you define "over Tribler the network". The nuance is that "traffic over I2P" will go through our network, not "I2P traffic". Assuming you meant the former: yes, correct.

**Comment by teddyrogers** [comment]
> It depends on how you define "over Tribler the network". The nuance is that "traffic over I2P" will go through our network, not "I2P traffic".

Through the Tribler network is indeed what I meant.

How will I2P work in practice, will we be able to designate the number of Tribler hops, before Tribler connects to an I2P entry node, then traffic is routed over I2P gateway as normal?

**Comment by qstokkink** [comment]
@teddyrogers That is correct. Your experience should pretty much be the same, except that you might see some I2P peers appearing in the peer list of your downloads.

**Comment by teddyrogers** [comment]
@qstokkink good to know!

Some more questions regarding I2P implementation, will we be able to...

1. ...add and use I2P trackers? E.g. http://w7tpbzncbcocrqtwwm3nezhnnsw4ozadvi2hmvzdhrqzfxfum7wa.b32.i2p/a
2. ...define the number of inbound and outbound [I2P tunnels including hops](https://geti2p.net/en/docs/how/tunnel-routing)?

**Comment by qstokkink** [comment]
"No" to (1) and (2): for both of those features, we need to connect to/through a local I2P service, which would be able to set those things.

That also relates to my previous answer. Those features are related to running "I2P traffic", instead of "traffic over I2P".

**Comment by teddyrogers** [comment]
@qstokkink how many default I2P inbound and outbound tunnels and hops will Tribler be utilising?

Can we utilise a (socks) connection to connect to I2P http websites? While on the subject can this be done in current Tribler network, on 8.0.7?
 
I know this is a subjective question and answer, what are the I2P speeds like?

**Comment by qstokkink** [comment]
Tribler is, in the next release (not 8.0.7), able to _connect_ to the edges of the I2P network. It doesn't have "I2P inbound and outbound tunnels and hops". You cannot "connect to I2P http eepsites". For those things, you'll have to actually use I2P instead of Tribler.

> what are the I2P speeds like?

From what I read on the `libtorrent` GitHub, speeds offered by I2P peers are so low that they are usually kicked from the peer list in favor of "vanilla" Bittorrent-protocol peers. So, this functionality will probably only be noticable when you are downloading "rare" torrents and there are no other peers available.

**Comment by absolutep** [comment]
> > will Tribler support I2P DHT or I2P PeX?
> 
> This is not supported by our [underlying library](https://github.com/arvidn/libtorrent/issues/7408). That means this is a bit more effort than "turn feature on". So, maybe in the future. But, not short term. 

I2P PeX @qstokkink  is being tested/implemented right now https://github.com/arvidn/libtorrent/pull/7831


> We would have to first investigate the protocol, see if it fits our implementation, revise our software architecture, etc. etc.

Yes, I understand, just attaching it here so that you can revisit or understand or coordinate with libtorrent developer on this if you have any coding related questions, cause of difference in approach of anonymous architecture etc.

Thanks.

**Comment by teddyrogers** [comment]
> Tribler is, in the next release (not 8.0.7)

Apologies, what I meant was in Tribler network, as it is now in 8.0.7 (not I2P related), would it be possible to use for http requests to access websites?

> It doesn't have "I2P inbound and outbound tunnels and hops".

Dumb it down for me, when Tribler is connected to I2P it looks similar to this?

Tribler Client -> Tribler Network -> SAM Bridge -> I2P Network

**Comment by qstokkink** [comment]
> would it be possible to use for http requests to access websites?

Ah, I see: no, this is blocked by Tribler's exit node script (assuming nobody runs a modified exit node script). Only Bittorrent related data passes through, like http tracker requests.

> Tribler Client -> Tribler Network -> SAM Bridge -> I2P Network

AFAIK, with the disclaimer that there are a lot of changes being made as we speak (see also @absolutep's most recent comment): we now allow incoming `Tribler Client <-> Tribler Network <- I2P Network`. If we hook in some additional settings, we could get `Tribler Client <-> Tribler Network <-> SAM Bridge <-> I2P Network`, but (again see @absolutep's comment) this seems like it doesn't work quite right yet in our underlying library.

**Comment by absolutep** [comment]
Please, reopen this as it is still not implemented

**Comment by qstokkink** [comment]
I2P peers have been enabled. This is implemented. You should be able to see them appear in downloads.

I2P proxying has not been enabled. For this, please refer to #8432

I don't see a need to reopen this issue.
